Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,497 @@
+    (n.factorial : A) * hI.dpow n x = x ^ n := by
+  induction n with
+  | zero => rw [factorial_zero, cast_one, one_mul, pow_zero, hI.dpow_zero hx]
+  | succ n ih =>
+    rw [factorial_succ, mul_comm (n + 1)]
+    nth_rewrite 1 [← (n + 1).choose_one_right]
+    rw [← choose_symm_add, cast_mul, mul_assoc,
+      ← hI.mul_dpow n 1 hx, ← mul_assoc, ih, hI.dpow_one hx, pow_succ, mul_comm]
+
+theorem dpow_eval_zero {n : ℕ} (hn : n ≠ 0) : hI.dpow n 0 = 0 := by
+  rw [← MulZeroClass.mul_zero (0 : A), hI.dpow_mul n I.zero_mem,
+    zero_pow hn, zero_mul, zero_mul]
+
+/-- If an element of a divided power ideal is killed by multiplication
+by some nonzero integer `n`, then its `n`th power is zero.
+
+Proposition 1.2.7 of [Berthelot-1974], part (i). -/
+theorem nilpotent_of_mem_dpIdeal {n : ℕ} (hn : n ≠ 0) (hnI : ∀ {y} (_ : y ∈ I), n • y = 0)
+    (hI : DividedPowers I) {x} (hx : x ∈ I) :
+    x ^ n = 0 := by
+  have h_fac : (n.factorial : A) * hI.dpow n x =
+    n • ((n - 1).factorial : A) * hI.dpow n x := by
+    rw [nsmul_eq_mul, ← cast_mul, mul_factorial_pred (Nat.pos_of_ne_zero hn)]
+  rw [← factorial_mul_dpow_eq_pow hI _ _ hx, h_fac, smul_mul_assoc]
+  exact hnI (I.mul_mem_left ((n - 1).factorial : A) (hI.dpow_mem hn hx))
+
+/-- If J is another ideal of A with divided powers,
+then the divided powers of I and J coincide on I • J
+
+[Berthelot-1974], 1.6.1 (ii) -/
+theorem coincide_on_smul {J : Ideal A} (hJ : DividedPowers J) {n : ℕ} {a : A} (ha : a ∈ I • J) :
+    hI.dpow n a = hJ.dpow n a := by
+  induction ha using Submodule.smul_induction_on' generalizing n with
+  | smul a ha b hb =>
+    rw [Algebra.id.smul_eq_mul, hJ.dpow_mul n hb, mul_comm a b, hI.dpow_mul n ha, ←
+      hJ.factorial_mul_dpow_eq_pow n b hb, ← hI.factorial_mul_dpow_eq_pow n a ha]
+    ring
+  | add x hx y hy hx' hy' =>
+    rw [hI.dpow_add n (Ideal.mul_le_right hx) (Ideal.mul_le_right hy),
+      hJ.dpow_add n (Ideal.mul_le_left hx) (Ideal.mul_le_left hy)]
+    apply sum_congr rfl
+    intro k _
+    rw [hx', hy']
+
+/-- A product of divided powers is a multinomial coefficient times the divided power
+
+[Roby-1965], formula (III') -/
+theorem prod_dpow {ι : Type*} {s : Finset ι} (n : ι → ℕ) {a : A} (ha : a ∈ I) :
+    (s.prod fun i => hI.dpow (n i) a) = multinomial s n * hI.dpow (s.sum n) a := by
+  classical
+  induction s using Finset.induction with
+  | empty =>
+    simp only [prod_empty, multinomial_empty, cast_one, sum_empty, one_mul]
+    rw [hI.dpow_zero ha]
+  | insert hi hrec =>
+    rw [prod_insert hi, hrec, ← mul_assoc, mul_comm (hI.dpow (n _) a),
+      mul_assoc, mul_dpow _ _ _ ha, ← sum_insert hi, ← mul_assoc]
+    apply congr_arg₂ _ _ rfl
+    rw [multinomial_insert hi, mul_comm, cast_mul, sum_insert hi]
+
+-- TODO : can probably be simplified using `DividedPowers.exp`
+
+/-- Lemma towards `dpow_sum` when we only have partial information on a divided power ideal
+
+See also `dpow_sum_aux'` for a more general result-/
+theorem dpow_sum_aux (dpow : ℕ → A → A)
```

Comment:
Maybe `private`?

Reply:
The lemma is there in hope that it will be useful elsewhere. Finding a better name would be more appropriate than `private`izing the definition.

Reply:
I made this lemma slightly more general, renamed it as `dpow_sum'` and deleted the second one (which is now mathematically equivalent to the new one, but complicated to use, because of coercions.)